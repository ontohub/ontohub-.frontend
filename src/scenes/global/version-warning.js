import React, { Component } from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { Message } from "semantic-ui-react";
import { satisfies } from "semver";
import styled from "styled-components";

const query = gql`
  query GetVersion {
    version {
      tag
      commitsSinceTag
    }
  }
`;

const BottomMessage = styled(Message)`
  position: fixed !important;
  margin: -1px !important;
  border-radius: 0 !important;
  bottom: 0;
  width: calc(100% + 2px) !important;
`;

const WarningMessage = ({ icon, warning, ...props } = {}) => (
  <BottomMessage
    negative={!warning}
    warning={!!warning}
    icon={icon || "warning sign"}
    {...props}
  />
);

const VersionWarning = ({ requirement }) => (
  <Query query={query}>
    {({ error, loading, data }) => {
      if (error) {
        return (
          <WarningMessage
            icon="plug"
            warning
            header="Could not determine backend version"
            content="This could mean that the backend is currently offline"
          />
        );
      } else if (loading) {
        return null;
      }
      const { tag, commitsSinceTag } = data.version;
      const version = `${tag}-${commitsSinceTag}`;
      const validVersion = satisfies(version, requirement);

      return validVersion ? null : (
        <WarningMessage
          header="The connected backend does not meet the version requirement"
          content={
            <p>
              Expected version <code>{version}</code> to satisfy requirement{" "}
              <code>{requirement}</code>
              . Be aware that this may cause problems.
            </p>
          }
        />
      );
    }}
  </Query>
);

export { VersionWarning, WarningMessage, query };
