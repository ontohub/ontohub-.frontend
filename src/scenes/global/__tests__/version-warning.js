import { WarningMessage, VersionWarning, query } from "../version-warning";

describe("WarningMessage", () => {
  set("component", () =>
    shallow(
      <WarningMessage header="Message header" content="Message content" />
    )
  );
  it("matches the snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});

describe("VersionWarning", () => {
  set("component", () => render(<VersionWarning requirement={requirement} />));
  set("requirement", () => "> v0.0.0-65");

  describe("version could not be determined", () => {
    it("shows an error message", () => {
      mockResponse(query, { error: true });
      expect(component).toMatchSnapshot();
    });
  });

  describe("version is being loaded", () => {
    it("shows no message", () => {
      mockResponse(query, { loading: true });
      expect(component).toMatchSnapshot();
    });
  });

  describe("successful version retrieval", () => {
    describe("version satisfies requirement", () => {
      set("tag", () => "v0.0.0");
      set("commitsSinceTag", () => 88);

      it("is hidden", () => {
        mockResponse(query, {
          data: { version: { tag, commitsSinceTag } }
        });
        expect(component).toMatchSnapshot();
      });
    });

    describe("version does not satisfy requirement", () => {
      set("tag", () => "v0.0.0");
      set("commitsSinceTag", () => 65);

      it("matches the snapshot", () => {
        mockResponse(query, {
          data: { version: { tag, commitsSinceTag } }
        });
        expect(component).toMatchSnapshot();
      });
    });
  });
});
