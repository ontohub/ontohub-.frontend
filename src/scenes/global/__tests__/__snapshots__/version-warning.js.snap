// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionWarning successful version retrieval version does not satisfy requirement matches the snapshot 1`] = `
.c0 {
  position: fixed !important;
  margin: -1px !important;
  border-radius: 0 !important;
  bottom: 0;
  width: calc(100% + 2px) !important;
}

<div
  class="ui icon negative message c0"
>
  <i
    aria-hidden="true"
    class="warning sign icon"
  >
    
  </i>
  <div
    class="content"
  >
    <div
      class="header"
    >
      The connected backend does not meet the version requirement
    </div>
    <p>
      Expected version 
      <code>
        v0.0.0-65
      </code>
       to satisfy requirement 
      <code>
        &gt; v0.0.0-65
      </code>
      . Be aware that this may cause problems.
    </p>
  </div>
</div>
`;

exports[`VersionWarning successful version retrieval version satisfies requirement is hidden 1`] = `null`;

exports[`VersionWarning version could not be determined shows an error message 1`] = `
.c0 {
  position: fixed !important;
  margin: -1px !important;
  border-radius: 0 !important;
  bottom: 0;
  width: calc(100% + 2px) !important;
}

<div
  class="ui icon warning message c0"
>
  <i
    aria-hidden="true"
    class="plug icon"
  >
    
  </i>
  <div
    class="content"
  >
    <div
      class="header"
    >
      Could not determine backend version
    </div>
    <p>
      This could mean that the backend is currently offline
    </p>
  </div>
</div>
`;

exports[`WarningMessage matches the snapshot 1`] = `
<Styled(Message)
  content="Message content"
  header="Message header"
  icon="warning sign"
  negative={true}
  warning={false}
/>
`;
